aws/resource_aws_sns_topic.go

```diff
@@ -142,13 +153,40 @@ func resourceAwsSnsTopicCreate(d *schema.ResourceData, meta interface{}) error {
 		name = resource.UniqueId()
 	}
 
+	fifoTopic := d.Get("fifo_topic").(bool)
+	cbd := d.Get("content_based_deduplication").(bool)
+
+	if fifoTopic {
+		if errors := validateSNSFifoTopicName(name); len(errors) > 0 {
+			return fmt.Errorf("Error validating the SNS FIFO topic name: %v", errors)
+		}
+	} else {
+		if errors := validateSNSNonFifoTopicName(name); len(errors) > 0 {
+			return fmt.Errorf("Error validating SNS topic name: %v", errors)
+		}
+	}
+
+	if !fifoTopic && cbd {
+		return fmt.Errorf("Content based deduplication can only be set with FIFO topics")
+	}
```

Comment:
You can replace the validations during create (i.e. `terraform apply`) with validations during plan (`terraform plan`) by adding a `CustomizeDiff` function to the resource:

```go
func resourceAwsSnsTopic() *schema.Resource {
	return &schema.Resource{
		Create: resourceAwsSnsTopicCreate,
		Read:   resourceAwsSnsTopicRead,
		Update: resourceAwsSnsTopicUpdate,
		Delete: resourceAwsSnsTopicDelete,
		Importer: &schema.ResourceImporter{
			State: schema.ImportStatePassthrough,
		},

		CustomizeDiff: resourceAwsSnsTopicCustomizeDiff,

		Schema: map[string]*schema.Schema{
...
}

func resourceAwsSnsTopicCustomizeDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
	fifoTopic := diff.Get("fifo_topic").(bool)
	contentBasedDeduplication := diff.Get("content_based_deduplication").(bool)

	if diff.Id() == "" {
		// Create.

		name := naming.Generate(diff.Get("name").(string), diff.Get("name_prefix").(string))

		if fifoTopic {
			if errors := validateSNSFifoTopicName(name); len(errors) > 0 {
				return fmt.Errorf("Error validating the SNS FIFO topic name: %v", errors)
			}
		} else {
			if errors := validateSNSNonFifoTopicName(name); len(errors) > 0 {
				return fmt.Errorf("Error validating SNS topic name: %v", errors)
			}
		}
	}

	if !fifoTopic && contentBasedDeduplication {
		return fmt.Errorf("Content based deduplication can only be set with FIFO topics")
	}

	return nil
}
```

Reply:
See also [this comment](https://github.com/hashicorp/terraform-provider-aws/pull/17164#issuecomment-762856206) around using the `naming` internal package and adding a `GenerateWithSuffix` function.

Reply:
https://github.com/hashicorp/terraform-provider-aws/pull/15828/commits/95fd6beffc1b52e58be7bed7d131a0270e06fb1c adds the `CustomizeDiff` function.

Just double checking here, `naming.GenerateWithSuffix` doesn't currently exist (per [the docs](https://pkg.go.dev/github.com/terraform-providers/terraform-provider-aws@v1.60.0/aws)) and shouldn't be part of this PR?

Reply:
@tomekr You can add. Thanks.

Reply:
@ewbankkit My apologies, I'm not quite following where I should add `naming.GenerateWithSuffix`. https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/contribution-checklists.md#resource-name-generation-code-implementation doesn't appear to mention adding a new function to `naming`.

Reply:
@ewbankkit  just following back up on this thread

Reply:
@ewbankkit trying one more time here